code/src/i2c_bus.cpp

```diff
@@ -0,0 +1,148 @@
+#include <i2c_bus.hpp>
+
+namespace r2d2::i2c {
+    void i2c_bus_c::pin_init() {
+        auto config_pin = [](uint32_t pin) {
+            PIOA->PIO_ABSR &= (~pin & PIOA->PIO_ABSR);
+            PIOA->PIO_PDR = pin;
+            PIOA->PIO_IDR = pin;
+            PIOA->PIO_PUER = pin;
+        };
+
+        if (_selected == TWI0) {
+            config_pin(PIO_PA18A_TWCK0);
+            config_pin(PIO_PA17A_TWD0);
+        } else if (_selected == TWI1) {
+            config_pin(PIO_PB13A_TWCK1);
+            config_pin(PIO_PB12A_TWD1);
+        }
+    }
+    void i2c_bus_c::clock_init() {
+        if (_selected == TWI0) {
+            if ((PMC->PMC_PCSR0 & (1 << ID_TWI0)) != (1 << ID_TWI0)) {
+                PMC->PMC_PCER0 |= 1 << ID_TWI0;
+            }
+        } else if (_selected == TWI1) {
+            if ((PMC->PMC_PCSR0 & (1 << ID_TWI1)) != (1 << ID_TWI1)) {
+                PMC->PMC_PCER0 |= 1 << ID_TWI1;
+            }
+        }
+
+        _selected->TWI_CR |= TWI_CR_SVEN;
+        _selected->TWI_CR |= TWI_CR_SWRST;
+        _selected->TWI_CR |= TWI_CR_SVDIS;
+        _selected->TWI_CR |= TWI_CR_MSDIS;
+        _selected->TWI_CR |= TWI_CR_MSEN;
+
+        static constexpr uint32_t masterClock =
+            84000000;        ///< Master clock rate 84MHz
+        uint32_t ckdiv = 0;  ///< Clock divider
+        uint32_t cLHDiv = 0; ///< Clock low and high divider
+        if (SPEED > 400000) {
+            return;
+        }
+        cLHDiv = masterClock / (SPEED * 2) - 4;
+
+        while ((cLHDiv > 0xFF) &&
+               (ckdiv <
+                7)) { ///< cldiv must fit in 8 bits, ckdiv must fit in 3 bits
+            ckdiv++;  ///< Increase clock devider
+            cLHDiv /= 2;
+        }
+        _selected->TWI_CWGR |= TWI_CWGR_CLDIV(cLHDiv) | TWI_CWGR_CHDIV(cLHDiv) |
+                               TWI_CWGR_CKDIV(ckdiv);
+    }
+    void i2c_bus_c::init() {
+        static bool init_done = false;
+        if (init_done) {
+            return;
+        }
+        if (_selected == nullptr) {
+            return;
+        }
+        pin_init();
+        clock_init();
+        init_done = true;
+    }
+    void i2c_bus_c::write_byte(const uint8_t data) {
+        _selected->TWI_THR = data;
+    }
+    uint8_t i2c_bus_c::read_byte() {
+        return _selected->TWI_RHR;
+    }
+    i2c_bus_c::i2c_bus_c(const interface &selected_interface,
+                         const uint32_t &SPEED)
+        : SPEED(SPEED) {
+        if (selected_interface == interface::interface_0) {
+            _selected = TWI0;
+        } else if (selected_interface == interface::interface_1) {
+            _selected = TWI1;
+        }
+        init();
+    }
+    void i2c_bus_c::write(const uint_fast8_t address, const uint8_t data[],
+                          const size_t &n) {
```

Comment:
You don't need to pass a size_t by ref

Reply:
See previous comment to the "you don't have to pass a uint32_t by ref"